clustered/client/src/main/java/org/ehcache/clustered/client/internal/store/CommonServerStoreProxy.java

```diff
@@ -177,4 +184,70 @@ public void clear() throws TimeoutException {
       throw new ServerStoreProxyException(e);
     }
   }
+
+  @Override
+  public Iterator<Chain> iterator() throws TimeoutException {
+    EhcacheEntityResponse.IteratorBatch iteratorBatch = openIterator();
+    if (iteratorBatch.isTerminated()) {
+      return iteratorBatch.getChains().iterator();
+    } else {
+      UUID iteratorId = iteratorBatch.getIdentity();
+      return new Iterator<Chain>() {
+
+        private boolean terminalBatch = false;
+        private Iterator<Chain> batch = iteratorBatch.getChains().iterator();
+
+        @Override
+        public boolean hasNext() {
+          return !terminalBatch || batch.hasNext();
+        }
+
+        @Override
+        public Chain next() {
+          if (batch.hasNext()) {
+            return batch.next();
+          } else {
+            try {
+              EhcacheEntityResponse.IteratorBatch batchResponse = fetchBatch(iteratorId);
+              batch = batchResponse.getChains().iterator();
+              terminalBatch = batchResponse.isTerminated();
+              return batch.next();
+            } catch (TimeoutException e) {
+              throw new RuntimeException(e);
+            }
+          }
+        }
+
+        @Override
+        protected void finalize() throws Throwable {
+          entity.invokeAndWaitForReceive(new ServerStoreOpMessage.IteratorCloseMessage(iteratorId), false);
```

Comment:
I think you can avoid this if `terminalBatch` is set.

Reply:
Makes sense.